mypy/subtypes.py

```diff
@@ -286,45 +303,169 @@ def is_callable_subtype(left: CallableType, right: CallableType,
+            # name-only arg, merge them.
+            if (left_by_position.required is False and left_by_name.required is False
+                    and left_by_position.name is None and left_by_name.pos is None
+                    and is_equivalent(left_by_position.typ, left_by_name.typ)):
+                left_arg = FormalArgument(
+                    left_by_name.name,
+                    left_by_position.pos,
+                    left_by_position.typ,
+                    required=False)
+            else:
+                return False
+        else:
+            left_arg = left_by_name if left_by_name is not None else left_by_position
 
-def is_var_arg_callable_subtype_helper(left: CallableType, right: CallableType) -> bool:
-    """Is left a subtype of right, assuming left has *args?
+        assert left_arg is not None
 
-    See also is_callable_subtype for additional assumptions we can make.
-    """
-    left_fixed = left.max_fixed_args()
-    right_fixed = right.max_fixed_args()
-    num_fixed_matching = min(left_fixed, right_fixed)
-    for i in range(num_fixed_matching):
-        if not is_subtype(right.arg_types[i], left.arg_types[i]):
+        if not is_left_more_general(left_arg, right_arg, ignore_pos_arg_names):
             return False
-    if not right.is_var_arg:
-        for i in range(num_fixed_matching, len(right.arg_types)):
-            if not is_subtype(right.arg_types[i], left.arg_types[-1]):
+
+    for i in range(len(left.arg_types)):
+        left_name = left.arg_names[i]
+        left_kind = left.arg_kinds[i]
+        left_type = left.arg_types[i]
+        # All *required* left-hand arguments must have a corresponding
+        # right-hand argument.  The relationship between the two is handled in
+        # the loop above.
+
+        if left_kind == ARG_POS and right.argument_by_position(i) is None:
+            if left_name is None:
                 return False
-        return True
-    else:
-        for i in range(left_fixed, right_fixed):
-            if not is_subtype(right.arg_types[i], left.arg_types[-1]):
+            if right.argument_by_name(left_name) is None:
+                return False
+        elif left_kind == ARG_NAMED:
+            if right.argument_by_name(left_name) is None:
+                return False
+        # Also check that *args and **kwargs types match in this loop
+        elif left_kind == ARG_STAR:
+            if right_star_type is not None and not is_subtype(right_star_type, left_type):
                 return False
-        for i in range(right_fixed, left_fixed):
-            if not is_subtype(right.arg_types[-1], left.arg_types[i]):
+        elif left_kind == ARG_STAR2:
+            if right_star2_type is not None and not is_subtype(right_star2_type, left_type):
                 return False
-        return is_subtype(right.arg_types[-1], left.arg_types[-1])
+    return True
+
+
+def is_left_more_general(
```

Comment:
I wonder if this could have a more descriptive name, since this also checks things that arguably aren't about being more general, such as position. Random ideas: `are_args_compatible`, `is_left_corresponding_and_more_general`